### PR TITLE
PP 4439 - Refactor PaymentProvider.refund() response

### PR DESCRIPTION
**Pull request by kbottla**
## WHAT
- This change simplifies the response object (GatewayResponse<BaseRefundResponse> --> GatewayRefundResponse) returned from PaymentProvider refund operation (inline with capture refactoring PR #898).
- ChargeRefundService has also been refactored to update refund based on state returned by PaymentProvider and not based on specific payment provider

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/gateway/model/response/GatewayRefundResponse.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    public Optional<GatewayError> getError() {
+        return Optional.ofNullable(gatewayError);
+    }
+
+    public GatewayRefundResponse.RefundState state() {
+        return refundState;
+    }
+
+    public Optional<String> getReference() {
+        return reference;
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/gateway/model/response/GatewayRefundResponse.java:45]: this should be `return Optional.ofNullable(reference)` as you're passing in a null reference in line 26. also if there's an error it's not guaranteed the `refundResponse.getReference()` is not null (line 31).

**Review by oswaldquek — CHANGES_REQUESTED** [review]
haven't gotten round to reviewing everything yet but just one comment so far

File: src/main/java/uk/gov/pay/connector/gateway/model/response/GatewayRefundResponse.java (hunk 1 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/model/response/GatewayRefundResponse.java:45]: 👍 Good spot. I have changed factory method to pass `Optional.empty()` instead of null. As reference is already Optional, can't use `return Optional.ofNulllable` straight away

File: src/main/java/uk/gov/pay/connector/gateway/stripe/handler/StripeRefundHandler.java [hunk 2]
```diff
@@ -58,28 +55,28 @@ public StripeRefundHandler(StripeGatewayClient client, StripeGatewayConfig strip
...
-            logger.error("Refund failed for transaction id {}. Failure code from Stripe: {}, failure message from Stripe: {}. ErrorId: {}. Response code from Stripe: {}",
-                    request.getTransactionId(), stripeErrorResponse.getError().getCode(), stripeErrorResponse.getError().getMessage(), errorId, response.getStatus());
-            GatewayError gatewayError = genericGatewayError(stripeErrorResponse.getError().getMessage());
-            return responseBuilder.withGatewayError(gatewayError).build();
+            StripeErrorResponse.Error error = stripeErrorResponse.getError();
+            logger.error("Refund failed for transaction id {}. Failure code from Stripe: {}, failure message from Stripe: {}. Response code from Stripe: {}",
+                    request.getTransactionId(), error.getCode(), error.getMessage(), response.getStatus());
 
+            return fromBaseRefundResponse(
+                    StripeRefundResponse.of(request.getTransactionId(), error.getCode(), error.getMessage()),
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/handler/StripeRefundHandler.java]: I don't think passing in the `request.getTransactionId()` as the reference is correct here. The reference is an id returned by Stripe as a result of a successful refund request. The `request.getTransactionId()` is actually the charge entity's gateway transaction id: https://github.com/alphagov/pay-connector/blob/32a0acca2b169c9c4ca34d2e8d54bcf45f293ba4/src/main/java/uk/gov/pay/connector/gateway/model/request/RefundGatewayRequest.java#L59 and is not the reference.

since `GatewayClientException` is a 4xx error and Stripe seems to return something like
```
{
  "error": {
    "message": "The transfer tr_blah_blah_blah is already fully reversed.",
    "type": "invalid_request_error"
  }
}
```
 i think the reference should be null in this case.

File: src/main/java/uk/gov/pay/connector/gateway/stripe/response/StripeRefundResponse.java [hunk 3]
```diff
@@ -15,16 +15,45 @@ private StripeRefundResponse(String reference) {
...
-    public static StripeRefundResponse of(Response response) {
-        final String id = response.readEntity(Map.class).get("id").toString();
-        return new StripeRefundResponse(id);
+    private StripeRefundResponse(String reference, String errorCode, String errorMessage) {
+        this.reference = reference;
+        this.errorCode = errorCode;
+        this.errorMessage = errorMessage;
+    }
+
+    public static StripeRefundResponse of(String reference, String errorCode, String errorMessage) {
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/response/StripeRefundResponse.java]: since reference might be null in the event of an error it's likely you'll need a constructor that only takes an errorCode and errorMessage

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java [hunk 4]
```diff
@@ -104,34 +104,47 @@ public RefundEntity createRefund(Long accountId, String chargeId, RefundRequest
...
 
-            refundEntity.setStatus(status);
-            getRefundReference(refundEntity, gatewayResponse).ifPresent(refundEntity::setReference);
+            if (refundStatus == REFUNDED) {
+                refundEntity.setStatus(REFUND_SUBMITTED);
+            }
+            if (chargeEntity.getPaymentGatewayName() == PaymentGatewayName.SANDBOX
+                    && refundEntity.hasStatus(RefundStatus.REFUND_SUBMITTED)) {
+                userNotificationService.sendRefundIssuedEmail(refundEntity);
+            }
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: lines 121-127. one of the purposes of https://github.com/alphagov/pay-connector/pull/898 really is to make the `CardCaptureService` payment provider agnostic. i think you can do the same here instead of relying on the payment provider name? we should do this because Stripe and Sandbox payment providers both return a COMPLETE status, and therefore a REFUNDED status here.

Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: see https://github.com/alphagov/pay-connector/blob/beea6fa714eaf76f1531a536cabf2469d6484a18/src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java#L344-L355 as an example of making it payment provider agnostic.

**Review by oswaldquek — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/pay/connector/gateway/model/response/GatewayRefundResponse.java (hunk 1 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/gateway/model/response/GatewayRefundResponse.java:45]: 👍

File: src/main/java/uk/gov/pay/connector/gateway/stripe/handler/StripeRefundHandler.java (hunk 2 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/stripe/handler/StripeRefundHandler.java]: 👍 Removed reference

File: src/main/java/uk/gov/pay/connector/gateway/stripe/response/StripeRefundResponse.java (hunk 3 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/stripe/response/StripeRefundResponse.java]: Added new factory method for only errorCode and errorMessage

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java (hunk 4 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: As far the refund status is concerned, we are not depending anymore on payment provider and relying on refund state returned.
Only reason we still see payment provider and can't extract entire update logic is because of email notifications sent only for sandbox.

Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: I don't think sending a notification is a sandbox specific thing. It seems logical to me that if a refund's status is REFUNDED, the user should expect an email notification, regardless of whether the PSP was Stripe or Sandbox. but check with @mrlumbu or @alexbishop1

Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: Also I think lines 121-123 
```
if (refundStatus == REFUNDED) {
             getRefundReference(refundEntity, gatewayResponse).ifPresent(refundEntity::setReference);	                 refundEntity.setStatus(REFUND_SUBMITTED);
             }
```
is very confusing. The question is why is this being done??

Generally this section of code needs a bit more clarity.

File: src/main/java/uk/gov/pay/connector/gateway/stripe/response/StripeRefundResponse.java (hunk 3 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/gateway/stripe/response/StripeRefundResponse.java]: 👍 
`public static StripeRefundResponse of(String reference, String errorCode, String errorMessage)` isn't being used anymore though so can be removed.

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java (hunk 4 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: This is not a new functionality introduced.

https://github.com/alphagov/pay-connector/blob/master/src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java#L128-L132

Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: Not sure how lines 121-123 became
`if (refundStatus == REFUNDED) {
             getRefundReference(refundEntity, gatewayResponse).ifPresent(refundEntity::setReference);	                 refundEntity.setStatus(REFUND_SUBMITTED);
             }`

This is the code that I can see
`if (refundStatus == REFUNDED) {
                refundEntity.setStatus(REFUND_SUBMITTED);
            }`

Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: Sorry, not sure how I copied that wrong. but `if (refundStatus == REFUNDED) { refundEntity.setStatus(REFUND_SUBMITTED); }` is very confusing. why is this being done? there doesn't seem to be any rationale behind this code.

File: src/main/java/uk/gov/pay/connector/gateway/stripe/response/StripeRefundResponse.java (hunk 3 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/stripe/response/StripeRefundResponse.java]: Oops! Removed unused factory method

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java (hunk 4 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: "This is not a new functionality introduced."

Yes, of course. But we've introduced a new payment provider Stripe. It makes sense to send email notifications for any provider which returns a COMPLETE status.

Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: Depends on what our functionality is. 
- Do we send refund email notifications to users (at any point) ? cc @alexbishop1 @mrlumbu.

As far as I can see it is only for Sandbox that we send notifications.

Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: > if (refundStatus == REFUNDED) { refundEntity.setStatus(REFUND_SUBMITTED); } is very confusing. why is this being done? there doesn't seem to be any rationale behind this code.

This will enable to store history of events for refunds and also displayed (for services) in frontend as:
```
Refunded                           amount   date 
Refund Submitted by [user email]   amount   date 
```

Comment (mrlumbu) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: @kbottla no, we definitely send them for all gateways, it's just that I believe we send them once we have received a successful refund notification for ePDQ, Worldpay, Smartpay, which is why we have that hacky bit of code for sandbox. I can dig out the code in a bit, or I'm sure you can find it

Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: We don't at the moment. It is been missed out on refactoring notifications. Just raised another PR for the fix

Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: @kbottla the other PR is for worldpay. i believe we should send refund notification for Stripe which means we can get rid of references to the Sandbox payment provider here.

Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: basically, `if (refundStatus == REFUNDED)`, send refund notification.

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java [hunk 5]
```diff
@@ -104,34 +103,44 @@ public RefundEntity createRefund(Long accountId, String chargeId, RefundRequest
...
                     refundEntity.getExternalId(), chargeEntity.getPaymentGatewayName().getName(), refundEntity.getReference(),
                     chargeEntity.getExternalId(), chargeEntity.getPaymentGatewayName().getName(), chargeEntity.getGatewayTransactionId(),
                     chargeEntity.getGatewayAccount().getAnalyticsId(), chargeEntity.getGatewayAccount().getId(),
-                    gatewayResponse, refundEntity.getStatus(), status);
+                    gatewayRefundResponse, refundEntity.getStatus(), refundStatus);
 
-            refundEntity.setStatus(status);
-            getRefundReference(refundEntity, gatewayResponse).ifPresent(refundEntity::setReference);
+            if (refundStatus == REFUNDED) {
+                refundEntity.setStatus(REFUND_SUBMITTED);
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java]: i understand this is for storing history of events for refunds, but where is this history stored? 

if you look at 

https://github.com/alphagov/pay-connector/blob/beea6fa714eaf76f1531a536cabf2469d6484a18/src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java#L344-L355

it's very clear about this.

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java (hunk 4 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:135]: The other PR is for RefundNotificationsProcessor (which is a fix). I have already made suggested changes (`if (refundStatus == REFUNDED)`) for Stripe.

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java (hunk 5 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java]: RefundHistory is being handled by Customiser and eclipselink here
https://github.com/alphagov/pay-connector/blob/master/src/main/java/uk/gov/pay/connector/refund/model/domain/RefundEntity.java#L52

We can look at refactoring history writing separately if any improvements are needed.

Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java]: cool, that HistoryCustomizer looks interesting. i don't think any refactoring is needed. i'm more concerned about readability and understanding code - reading it as if i were a new developer on the project. i think there should be some comment then, about why you're setting the RefundEntity's status to REFUND_SUBMITTED. 
i think it'll also be good to modify SandboxRefundITest to assert the RefundHistory of a refund processed by Sandbox payment provider has a REFUND_SUBMITTED and REFUNDED status.

basically i think everything could be collapsed to:
```
if (refundStatus == REFUNDED) {
                refundEntity.setStatus(REFUND_SUBMITTED); //some comment about why we're doing this
                userNotificationService.sendRefundIssuedEmail(refundEntity);
                refundEntity.setStatus(REFUNDED);
            } else refundEntity.setStatus(refundStatus);
```

Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java]: sorry to be a pain

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java [hunk 6]
```diff
@@ -104,34 +103,47 @@ public RefundEntity createRefund(Long accountId, String chargeId, RefundRequest
...
                     refundEntity.getExternalId(), chargeEntity.getPaymentGatewayName().getName(), refundEntity.getReference(),
                     chargeEntity.getExternalId(), chargeEntity.getPaymentGatewayName().getName(), chargeEntity.getGatewayTransactionId(),
                     chargeEntity.getGatewayAccount().getAnalyticsId(), chargeEntity.getGatewayAccount().getId(),
-                    gatewayResponse, refundEntity.getStatus(), status);
+                    gatewayRefundResponse, refundEntity.getStatus(), refundStatus);
+
+            if (refundStatus == REFUNDED) {
+                // If the gateway confirms refunds immediately, insert history (REFUND_SUBMITTED) to record
+                // additional event for refund submitted date which also includes the user submitting the
+                // refund. This will also help services to view refund history in detail in self service
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java]: Can this be changed to ```If the gateway confirms refunds immediately, the refund status needs to be set to REFUND_SUBMITTED and then REFUNDED. This will help services to view refund history in detail in self service. By setting the refund status like `refundEntity.setStatus(REFUND_SUBMITTED)` for example, the refund status history is magically captured by the @Customizer(HistoryCustomizer.class) annotation on RefundEntity.```

Again I'm approaching this with the eyes of a new developer and I think it's important to tell HOW this history is persisted in the database, rather than saying "insert history to record".

**Review by oswaldquek — CHANGES_REQUESTED** [review]
sorry just one change requested to make the comment clearer for developers.

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java (hunk 6 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java]: I see your point. But I am not comfortable mixing up why (business desc) and how (tech choices) and also anytime (which is in multiple places) we use `refund.setStatus` \@Customiser comes into action. I feel this may not be right place to describe tech implementation

File: src/main/java/uk/gov/pay/connector/refund/model/domain/RefundHistory.java [hunk 7]
```diff
@@ -6,6 +6,11 @@
 import java.sql.Timestamp;
 import java.time.ZonedDateTime;
 
+/**
+ * Entity which represents Refunds history (a record for each event for a particular refund)
+ * Each time a new status is set for RefundEntity a new record is created in RefundsHistory (Refunds_History table in database)
+ * which happens magically by @Customizer(HistoryCustomizer.class) annotation on RefundEntity
+ */
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/refund/model/domain/RefundHistory.java:13]: we don't need this comment anymore. i originally thought it was the `@Customizer(HistoryCustomizer.class)` that updates the refunds_history table but it's actually the `@Transactional` annotation on a method that causes the refund entity and the history table to be updated.

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java [hunk 8]
```diff
@@ -101,6 +108,8 @@ public void shouldRespond_412_WhenSecondRefundRequestAmountAvailableMismatches()
         List<Map<String, Object>> refundsFoundByChargeId = databaseTestHelper.getRefundsByChargeId(defaultTestCharge.getChargeId());
         assertThat(refundsFoundByChargeId.size(), is(1));
         assertThat(refundsFoundByChargeId, hasItems(aRefundMatching(refundId, is(notNullValue()), defaultTestCharge.getChargeId(), refundAmount, "REFUNDED")));
+
+        assertRefundsHistoryInDBForSuccessfulOrPartialRefund(defaultTestCharge);
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java]: this can be deleted as it is already asserted in the `shouldBeAbleToRequestARefund_partialAmount` test which is sufficient. no need to go overboard here.

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java [hunk 9]
```diff
@@ -114,6 +123,8 @@ public void shouldBeAbleToRequestARefund_fullAmount() {
...
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java]: this can be deleted as it is already asserted in the `shouldBeAbleToRequestARefund_partialAmount` test which is sufficient. no need to go overboard here.

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java [hunk 10]
```diff
@@ -136,6 +147,8 @@ public void shouldBeAbleToRequestARefund_multiplePartialAmounts_andRefundShouldB
                 aRefundMatching(secondRefundId, is(notNullValue()), chargeId, secondRefundAmount, "REFUNDED"),
                 aRefundMatching(firstRefundId, is(notNullValue()), chargeId, firstRefundAmount, "REFUNDED")));
 
+        assertRefundsHistoryInDBForTwoRefunds(defaultTestCharge);
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java]: can delete this as already asserted in `shouldBeAbleToRefundTwoRequestsWhereAmountAvailableMatch`.

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java [hunk 11]
```diff
@@ -176,13 +189,15 @@ public void shouldFailRequestingARefund_whenChargeRefundIsFull() {
         postRefundFor(externalChargeId, refundAmount, defaultTestCharge.getAmount())
                 .statusCode(ACCEPTED.getStatusCode());
 
+        List<Map<String, Object>> refundsFoundByChargeId = databaseTestHelper.getRefundsByChargeId(chargeId);
+        assertThat(refundsFoundByChargeId.size(), is(1));
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java:193]: don't think we need this assertion.

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java [hunk 12]
```diff
@@ -176,13 +189,15 @@ public void shouldFailRequestingARefund_whenChargeRefundIsFull() {
...
+
+        assertRefundsHistoryInDBForSuccessfulOrPartialRefund(defaultTestCharge);
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java]: don't think we need this assertion.

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java [hunk 13]
```diff
@@ -196,6 +211,9 @@ public void shouldFailRequestingARefund_whenAmountIsBiggerThanChargeAmount() {
 
         List<Map<String, Object>> refundsFoundByChargeId = databaseTestHelper.getRefundsByChargeId(defaultTestCharge.getChargeId());
         assertThat(refundsFoundByChargeId.size(), is(0));
+
+        List<String> refundsHistory = databaseTestHelper.getRefundsHistoryByChargeId(defaultTestCharge.getChargeId()).stream().map(x -> x.get("status").toString()).collect(Collectors.toList());
+        assertThat(refundsHistory.size(), is(0));
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java:214]: this can be simplified by using `assertThat(refundDao.searchHistoryByChargeId(defaultTestCharge.getChargeId()), hasSize(0))`

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java [hunk 14]
```diff
@@ -210,6 +228,9 @@ public void shouldFailRequestingARefund_whenAmountIsBiggerThanAllowedChargeAmoun
 
         List<Map<String, Object>> refundsFoundByChargeId = databaseTestHelper.getRefundsByChargeId(defaultTestCharge.getChargeId());
         assertThat(refundsFoundByChargeId.size(), is(0));
+        
+        List<String> refundsHistory = databaseTestHelper.getRefundsHistoryByChargeId(defaultTestCharge.getChargeId()).stream().map(x -> x.get("status").toString()).collect(Collectors.toList());
+        assertThat(refundsHistory.size(), is(0));
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java:231]: don't need this assertion as already tested

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java [hunk 15]
```diff
@@ -224,6 +245,9 @@ public void shouldFailRequestingARefund_whenAmountIsLessThanOnePence() {
 
         List<Map<String, Object>> refundsFoundByChargeId = databaseTestHelper.getRefundsByChargeId(defaultTestCharge.getChargeId());
         assertThat(refundsFoundByChargeId.size(), is(0));
+
+        List<String> refundsHistory = databaseTestHelper.getRefundsHistoryByChargeId(defaultTestCharge.getChargeId()).stream().map(x -> x.get("status").toString()).collect(Collectors.toList());
+        assertThat(refundsHistory.size(), is(0));
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java:248]: don't need this assertion as already tested

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java [hunk 16]
```diff
@@ -246,6 +270,8 @@ public void shouldFailRequestingARefund_whenAPartialRefundMakesTotalRefundedAmou
         List<Map<String, Object>> refundsFoundByChargeId1 = databaseTestHelper.getRefundsByChargeId(defaultTestCharge.getChargeId());
         assertThat(refundsFoundByChargeId1.size(), is(1));
         assertThat(refundsFoundByChargeId1, hasItems(aRefundMatching(firstRefundId, is(notNullValue()), defaultTestCharge.getChargeId(), firstRefundAmount, "REFUNDED")));
+
+        assertRefundsHistoryInDBForSuccessfulOrPartialRefund(defaultTestCharge);
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java]: don't need this assertion as already tested

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java [hunk 17]
```diff
@@ -280,4 +306,16 @@ private String assertRefundResponseWith(Long refundAmount, ValidatableResponse v
 
         return refundId;
     }
+
+    private void assertRefundsHistoryInDBForSuccessfulOrPartialRefund(DatabaseFixtures.TestCharge defaultTestCharge) {
+        List<String> refundsHistory = databaseTestHelper.getRefundsHistoryByChargeId(defaultTestCharge.getChargeId()).stream().map(x -> x.get("status").toString()).collect(Collectors.toList());
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java:309]: i'd rather `refundDao.searchHistoryByChargeId` was used as you get the `List<RefundHistory>` returned. it will be quite easy to order the list based on historyStartDate. because i'm not sure asserting events "containsInAnyOrder" is actually correct here.

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java [hunk 18]
```diff
@@ -280,4 +306,16 @@ private String assertRefundResponseWith(Long refundAmount, ValidatableResponse v
...
+        assertThat(refundsHistory.size(), is(3));
+        assertThat(refundsHistory, containsInAnyOrder("REFUND SUBMITTED", "REFUNDED", "CREATED"));
+    }
+
+    private void assertRefundsHistoryInDBForTwoRefunds(DatabaseFixtures.TestCharge defaultTestCharge) {
+        List<String> refundsHistory = databaseTestHelper.getRefundsHistoryByChargeId(defaultTestCharge.getChargeId()).stream().map(x -> x.get("status").toString()).collect(Collectors.toList());
+        assertThat(refundsHistory.size(), is(6));
+        assertThat(refundsHistory, containsInAnyOrder("REFUND SUBMITTED", "REFUNDED", "CREATED", "REFUND SUBMITTED", "REFUNDED", "CREATED"));
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java]: same comment as above

**Review by oswaldquek — CHANGES_REQUESTED** [review]
i'm happy with the implementation. just two things: 1) testing the event history once is enough... we don't need to assert the event history for every test as it will further pollute existing tests and affect readability. in fact i'd prefer if you just created an assertEventHistoryForASuccessfulRefund test 2) it's better to use the refundDao to get event history as you can verify the order in which events happen as they have a timestamp which you can use to sort.

File: src/main/java/uk/gov/pay/connector/refund/model/domain/RefundHistory.java (hunk 7 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/refund/model/domain/RefundHistory.java:13]: It is still @Customizer(HistoryCustomizer.class) that is responsible to insert into history.

Transactional is more about enabling boundaries for the operations and if we do multiple operations in a single transaction they won't be treated individually.

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java (hunk 8 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java]: We do it refunds record. It is better to have refunds history together. 
Having only one test won't cover full/partial refunds and also if that tests gets removed we are back to square 1. I feel we should add them wherever we check for refunds.

File: src/main/java/uk/gov/pay/connector/refund/model/domain/RefundHistory.java (hunk 7 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/refund/model/domain/RefundHistory.java:13]: ok, i understand 👍

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java (hunk 8 above)
Comment (oswaldquek) [reply src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java]: ?? it's already tested in `shouldBeAbleToRequestARefund_partialAmount`. all i'm saying is you don't have to go overboard and re-test it here (and for nearly every method in this test class).

Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java]: May be it is confusing to have common method ?
My intention is, if each test is considered on its own it should cover `issue refund, check response, check refunds & refunds history in DB for that particular transaction`

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java (hunk 17 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java:309]: Not against using history_start_date. Updated query to order by statuses and removed  containsInAnyOrder.

**Comment by kbottla** [comment]
> 1) testing the event history once is enough... we don't need to assert the event history for every test as it will further pollute existing tests and affect readability. in fact i'd prefer if you just created an assertEventHistoryForASuccessfulRefund test 

I don't think it as a right place to test refundHistory alone as these are resource tests. As each test is for different scenario we verify `response, refund record from DB` in each test. I have extended tests to check for refunds history. Anyone looking at an individual test can understand different aspects to be looked out for these tests.

> 2) it's better to use the refundDao to get event history as you can verify the order in which events happen as they have a timestamp which you can use to sort.

Updated helper to sort by date and checking the statuses in order. Felt it a bit like polluting test class if mixing helper and dao for different reasons.

File: src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java [hunk 19]
```diff
@@ -176,13 +189,15 @@ public void shouldFailRequestingARefund_whenChargeRefundIsFull() {
         postRefundFor(externalChargeId, refundAmount, defaultTestCharge.getAmount())
                 .statusCode(ACCEPTED.getStatusCode());
 
+        List<Map<String, Object>> refundsFoundByChargeId = databaseTestHelper.getRefundsByChargeId(chargeId);
+        assertThat(refundsFoundByChargeId.size(), is(1));
+
+        assertRefundsHistoryInOrderInDBForSuccessfulOrPartialRefund(defaultTestCharge);
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java]: we definitely don't need this assertion here as this is already done in the `shouldBeAbleToRequestARefund_fullAmount` test. this test (`shouldFailRequestingARefund_whenChargeRefundIsFull`) makes a full refund as well before making another refund. so the event history is already tested.

Comment (oswaldquek) [reply src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java]: we should keep code as clean as possible and not add unnecessary tests

Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/it/resources/sandbox/SandboxRefundITest.java]: 👍 Removed.

**Review by oswaldquek — APPROVED** [review]

**Comment by oswaldquek** [comment]
I actually think we should test refund history as it's own test as the end user can see this history. This is got by calling the `"/v1/api/accounts/{accountId}/charges/{chargeId}/events"` endpoint. So it makes perfect sense to have a resource level test where a refund is made and we assert the event history. This PR has gone on for long enough though so will approve. I will create the test separately.

**Comment by kbottla** [comment]
There is probably a test already although I haven't looked into its details
https://github.com/alphagov/pay-connector/blob/master/src/test/java/uk/gov/pay/connector/it/resources/ChargeEventsResourceITest.java#L92
